reana_job_controller/schemas.py

```diff
@@ -42,6 +47,38 @@ class JobRequest(Schema):
     voms_proxy = fields.Bool(required=False)
     kubernetes_uid = fields.Int(required=False)
     kubernetes_memory_limit = fields.Str(required=False)
+    kubernetes_job_timeout = fields.Int(required=False, strict=True)
     unpacked_img = fields.Bool(required=False)
     htcondor_max_runtime = fields.Str(required=False)
     htcondor_accounting_group = fields.Str(required=False)
+
+    @validates("kubernetes_job_timeout")
+    def validate_kubernetes_job_timeout(self, value):  # noqa: D102
+        if value <= 0:
+            raise ValidationError("kubernetes_job_timeout must be greater than 0.")
+
+        try:
+            max_value = int(REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT)
+        except (ValueError, TypeError):
+            raise ValidationError(
+                "Max value for kubernetes_job_timeout is not an integer. "
+                f"Provided value is '{REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT}'."
+            )
+
+        if value > max_value:
+            raise ValidationError(
+                f"kubernetes_job_timeout exceeds maximum allowed value of {max_value} seconds."
+            )
+
+    @pre_load
```

Comment:
If I understood it correctly, [this decorator](https://marshmallow.readthedocs.io/en/stable/marshmallow.decorators.html#marshmallow.decorators.pre_load) affects all the schema fields, which is not ideal in terms of performance.

If it's called for all the fields, the name of the function is a bit misleading as it looks like it's only affecting `kubernetes_job_timeout`.

Reply:
You can have multiple methods with the `pre_load` decorator. Each method is executed once as it provides the whole dictionary to the function, [example from docs](https://marshmallow.readthedocs.io/en/stable/extending.html#raising-errors-in-pre-post-processor-methods). Don't think it is a performance issue, more like "is this validation approach makes sense" question.

I don't think the name is misleading. The only purpose of this function is to set the `kubernetes_job_timeout` parameter hence a name. 

But, I have some ideas I need to test to make it more clear. It would be good to find a clear, simple way for future parameters. Let me test for today and I will get back.

Reply:
You're right, I thought it was called once per field, but this function can actually affect any field, that's why I thought that the name could be misleading, as we could potentially have more logic inside it, not related to Kubernetes job timeout directly.

Anyway, as you said, this approach is not that clear and too _verbose_ (hence, harder to maintain) just to set some default value and perform a basic validation. Maybe is worth taking a look at the built-in validators. I'll read about your findings later.

Reply:
> name could be misleading, as we could potentially have more logic inside it, not related to Kubernetes job timeout directly.

You can create another `pre_load` function to handle another parameter. Don't think it is misleading.

> Maybe is worth taking a look at the built-in validators. 

It is not possible to use built-in validators unless we will validate the default values before in `config.py`, for example. Otherwise, we will need to check if the default values entered by admins are correct in the first place.

Reply:
Regarding findings, not much. Came up with this snippet:

```python
from typing import Union
from marshmallow import Schema, fields, ValidationError

REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT = "20"
REANA_KUBERNETES_JOBS_TIMEOUT_LIMIT = "10"


class JobRequest(Schema):
    """Job request model."""

    kubernetes_job_timeout = fields.Method(missing=REANA_KUBERNETES_JOBS_TIMEOUT_LIMIT,
                                           deserialize="get_kubernetes_job_timeout")

    @staticmethod
    def get_kubernetes_job_timeout(value: Union[str, int]) -> int:
        try:
            value = int(value)
        except (ValueError, TypeError):
            raise ValidationError(f"kubernetes_job_timeout must be an integer. Got: {value}")

        if value <= 0:
            raise ValidationError("kubernetes_job_timeout must be greater than 0.")

        try:
            max_value = int(REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT)
        except (ValueError, TypeError):
            raise ValidationError(
                "Max value for kubernetes_job_timeout is not an integer. "
                f"Provided value is '{REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT}'. "
                "Please contact the administrator."
            )

        if value > max_value:
            raise ValidationError(
                f"kubernetes_job_timeout exceeds maximum allowed value of {max_value} seconds."
            )

        return value


job_request_schema = JobRequest()
#data, errors = job_request_schema.load({"kubernetes_job_timeout": "10", "hello": "world"})
data, errors = job_request_schema.load({})
print(data)
print(errors)
```

Advantage: simpler, one function is performing validation
Disadvantage: 
- if the default job timeout value by admin is not entered correctly you will receive the same error as if the user entered it;
- if `None` is a default value for job timeout (`REANA_KUBERNETES_JOBS_TIMEOUT_LIMIT=None`), it will get passed along without calling the `get_kubernetes_job_timeout` method (a very unlikely event).

Reply:
Another possibility:

```python
class JobRequest(Schema):
    kubernetes_job_timeout = fields.Int(required=False,
                                        strict=True,
                                        allow_none=False)

    @pre_load
    def set_kubernetes_job_timeout(self, in_data, **kwargs):
        if "kubernetes_job_timeout" not in in_data:
            try:
                in_data["kubernetes_job_timeout"] = int(REANA_KUBERNETES_JOBS_TIMEOUT_LIMIT)
            except (ValueError, TypeError):
                raise ValidationError(
                    "Default value of kubernetes_job_timeout is not an integer. "
                    f"Provided value is '{REANA_KUBERNETES_JOBS_TIMEOUT_LIMIT}'. "
                    "Please contact the administrator."
                )

        job_timeout = in_data["kubernetes_job_timeout"]

        try:
            job_timeout = int(job_timeout)
        except (ValueError, TypeError):
            raise ValidationError(f"kubernetes_job_timeout must be an integer. Provided value is '{job_timeout}'.")

        if job_timeout <= 0:
            raise ValidationError(
                "kubernetes_job_timeout must be greater than 0."
                f"Provided value is {job_timeout}."
            )

        try:
            max_value = int(REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT)
        except (ValueError, TypeError):
            raise ValidationError(
                "Max value for kubernetes_job_timeout is not an integer. "
                f"Provided value is '{REANA_KUBERNETES_JOBS_MAX_USER_TIMEOUT_LIMIT}'. "
                "Please contact the administrator."
            )

        if job_timeout > max_value:
            raise ValidationError(
                f"kubernetes_job_timeout exceeds maximum allowed value of {max_value} seconds. "
                f"Provided value is {job_timeout} seconds."
            )

        in_data["kubernetes_job_timeout"] = job_timeout
        return in_data
```

Advantage: simple, one function is performing validation, clear steps, can be extended later
Disadvantage:

- `pre_load` is passing the whole schema there. Function name and purpose are clear but it is not intended use of `pre_load`, I guess.

Reply:
Open to other suggestions. I don't know how we can use built-in validation without verifying default values somewhere else before.

Reply:
I prefer the second option with `pre_load`. It is clear and can be extended later easily with suffixes if needed.

Reply:
One thing about the second way is that in case of validation error with k8s job timeout, `job_request_schema.load` will return `"_schema": ["k8s_job_timeout error message"]` like style message. This message will end up in workflow engine logs, but it is still clear that it is problem with the job timeout value. I would say it is okay for now.

In case not, I can fix that by using the key in the `ValidationError` parameter.

Reply:
I went with a second option, a single `pre_load` function.

Reply:
I think the second one will do :+1: Unsure about the `_schema` part in the logs for regular users... However, since we don't expose r-j-controller's API directly, we can always sanitise all the arguments upfront. And the first workflows are usually run by the cluster admins, so they could catch validation issues from configuration.

Reply:
`_schema` part can be a bit weird but the message will still be clear - `kubernetes_job_timeout exceeds the maximum allowed value of 120 seconds`.

*Side note:* before this PR the error messages from JobRequest schema were not even visible in workflow engines. The smaller commit fixes it.